### PR TITLE
feature(api): add external logging

### PR DESCRIPTION
**Pull request by fivetran-aleksandrboldyrev**
Add support for Log Service Management API (https://fivetran.com/docs/rest-api/log-service-management)
Add Metadata API and Roles API Data Sources

### CODE REVIEW

File: fivetran/resource_external_logging.go [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+				Default:     false,
+				Description: "Specifies whether the setup tests should be run automatically. The default value is TRUE.",
+			},
+			"config": resourceExternalLoggingSchemaConfig(),
+		},
+	}
+}
+
+func resourceExternalLoggingSchemaConfig() *schema.Schema {
+	return &schema.Schema{Type: schema.TypeList, Required: true, MaxItems: 1,
```
Comment (beevital) [inline fivetran/resource_external_logging.go:54]: It reminds me the same code from datasource =) Let's make it common?

**Review by beevital — COMMENTED** [review]
LGTM, but let's avoid code duplication in new resources.
We already have a huge debt with old resources/datasources to refactor..

**Comment by fivetran-aleksandrboldyrev** [comment]
Merge main - it's about fixing conflicts via Github UI

File: fivetran/fivetran_test.go [hunk 2]
```diff
@@ -139,6 +140,19 @@ func cleanupDestinations() {
...
+	groups, err := client.NewGroupsList().Do(context.Background())
+	if err != nil {
+		log.Fatal(err)
+	}
+	for _, group := range groups.Data.Items {
+		_, err := client.NewExternalLoggingDelete().ExternalLoggingId(group.ID).Do(context.Background())
+		if err != nil && err.Error() != "status code: 404; expected: 200" {
+			log.Fatal(err)
+		}
+	}
```
Comment (beevital) [inline fivetran/fivetran_test.go:153]:
```suggestion
	}
	if groups.Data.NextCursor != nil {
	   cleanupExternalLogging()
	}
```

File: fivetran/tests/mock/resource_external_logging_test.go [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+        "service":            "azure_monitor_log",
+        "enabled":            false,
+        "config":{
+            "workspace_id":   "workspace_id",
+            "primary_key":    "primary_key",
+            "role_arn":       "role_arn",
+            "region":         "region",
+            "log_group_name": "log_group_name",
+            "sub_domain":     "sub_domain",
+            "enable_ssl":     "enable_ssl",
```
Comment (beevital) [inline fivetran/tests/mock/resource_external_logging_test.go:39]: shouldn't it be boolean?
Same for go-fivetran - it is a bug. It's won't work with real API.
```suggestion
            "enable_ssl":     true,
```

Comment (beevital) [reply fivetran/tests/mock/resource_external_logging_test.go:39]: public class SplunkLogCredentials implements ApiVerifiable {
    @ConfigField(type = FieldType.CONFIG_SECRED)
    public boolean enableSSL = true;

File: fivetran/resource_external_logging.go [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+				"channel": {
+					Type:        schema.TypeString,
+					Optional:    true,
+					Description: "Channel",
+				},
+				"enable_ssl": {
+					Type:        schema.TypeString,
+					Optional:    true,
+					Description: "Enable SSL",
+				},
```
Comment (beevital) [inline fivetran/resource_external_logging.go:141]:
```suggestion
				"enable_ssl": {
					Type:        schema.TypeBool,
					Optional:    true,
					Description: "Enable SSL",
				},
```

File: fivetran/resource_external_logging.go [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+		hasConfig = true
+	}
+	if v := c["hostname"].(string); v != "" {
+		fivetranConfig.Hostname(v)
+		hasConfig = true
+	}
+	if v := c["enable_ssl"].(string); v != "" {
+		fivetranConfig.EnableSsl(v)
+		hasConfig = true
+	}
```
Comment (beevital) [inline fivetran/resource_external_logging.go:328]:
```suggestion
	if v := c["enable_ssl"].(bool); v != "" {
		fivetranConfig.EnableSsl(v) // here will be a bug currently - we need to fix go-fivetran client. Or just use custom config.
		hasConfig = true
	}
```

File: fivetran/resource_external_logging.go [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+	return diags
+}
+
+// resourceExternalLoggingCreateConfig receives a config type []interface{} and returns a
+// *fivetran.ExternalLoggingConfig and a ok value. The ok value is true if any configuration
+// has been set.
+func resourceExternalLoggingCreateConfig(config []interface{}) (*fivetran.ExternalLoggingConfig, bool) {
+	fivetranConfig := fivetran.NewExternalLoggingConfig()
+	var hasConfig bool
+
```
Comment (beevital) [inline fivetran/resource_external_logging.go:286]: It's better to use map[string]interface{} and use DoCustom instead of Do request.

**Review by beevital — CHANGES_REQUESTED** [review]
We have a bug in go-fivetran - one field should be boolean.

Also terraform should use DoCustom request and represent config as a map.

File: fivetran/resource_external_logging.go (hunk 6 above)
Comment (fivetran-aleksandrboldyrev) [reply fivetran/resource_external_logging.go]: Done

File: fivetran/resource_external_logging.go [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+func resourceExternalLoggingCreateConfig(d *schema.ResourceData) map[string]interface{} {
+	configMap := make(map[string]interface{})
+
+	var config = d.Get("config").([]interface{})
+
+	if len(config) < 1 || config[0] == nil {
+		return configMap
+	}
+
+	c := config[0].(map[string]interface{})
```
Comment (beevital) [inline fivetran/resource_external_logging.go:283]: If we will have list-field - we will have to transform it, becayse in resource data it represented as (*schema.List) or (*schema.Set)
And it won't be correctly handled - as it is a pointer.
It's better to implement method that copies fields from resource data to request map one by one, verifying field type in schema and transform value if needed.

**Review by beevital — APPROVED** [review]
LGMT, have one suggestion in comment.

File: fivetran/resource_external_logging.go (hunk 7 above)
Comment (fivetran-aleksandrboldyrev) [reply fivetran/resource_external_logging.go:283]: Created follow-up task for this suggestion

**Pull request merged**